<a href="https://colab.research.google.com/github/rena-ds/manipulation-custchurn/blob/main/Data_Manipulation_on_Telco_Customer_Churn_Dataset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [24]:
#Import libraries
import pandas as pd
import os
import zipfile
import numpy as np

In [25]:
#Set directory for Kaggle file
os.environ['KAGGLE_CONFIG_DIR'] = os.path.expanduser('~/.kaggle')

#Command to download dataset
!kaggle datasets download -d "blastchar/telco-customer-churn"

#Extract ZIP file with zipfile
with zipfile.ZipFile("telco-customer-churn.zip", 'r') as zip_ref:
    zip_ref.extractall()

#Read dataset file(CSV)
df = pd.read_csv("WA_Fn-UseC_-Telco-Customer-Churn.csv")

Dataset URL: https://www.kaggle.com/datasets/blastchar/telco-customer-churn
License(s): copyright-authors
telco-customer-churn.zip: Skipping, found more recently modified local copy (use --force to force download)


In [26]:
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [37]:
df.shape

(7043, 21)

In [38]:
df.size

147903

In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [39]:
df.describe()

SeniorCitizen       tenure  MonthlyCharges
count    7043.000000  7043.000000     7043.000000
mean        0.162147    32.371149       64.761692
std         0.368612    24.559481       30.090047
min         0.000000     0.000000       18.250000
25%         0.000000     9.000000       35.500000
50%         0.000000    29.000000       70.350000
75%         0.000000    55.000000       89.850000
max         1.000000    72.000000      118.750000

In [40]:
df.describe(include = 'object')

customerID gender Partner Dependents PhoneService MultipleLines  \
count         7043   7043    7043       7043         7043          7043   
unique        7043      2       2          2            2             3   
top     7590-VHVEG   Male      No         No          Yes            No   
freq             1   3555    3641       4933         6361          3390   

       InternetService OnlineSecurity OnlineBackup DeviceProtection  \
count             7043           7043         7043             7043   
unique               3              3            3                3   
top        Fiber optic             No           No               No   
freq              3096           3498         3088             3095   

       TechSupport StreamingTV StreamingMovies        Contract  \
count         7043        7043            7043            7043   
unique           3           3               3               3   
top             No          No              No  Month-to-month   
freq          3473        2810            2785            3875   

       PaperlessBilling     PaymentMethod TotalCharges Churn  
count              7043              7043         7043  7043  
unique                2                 4         6531     2  
top                 Yes  Electronic check                 No  
freq               4171              2365           11  5174

In [41]:
df.dtypes

customerID           object
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges         object
Churn                object
dtype: object

In [42]:
df.nunique()

customerID          7043
gender                 2
SeniorCitizen          2
Partner                2
Dependents             2
tenure                73
PhoneService           2
MultipleLines          3
InternetService        3
OnlineSecurity         3
OnlineBackup           3
DeviceProtection       3
TechSupport            3
StreamingTV            3
StreamingMovies        3
Contract               3
PaperlessBilling       2
PaymentMethod          4
MonthlyCharges      1585
TotalCharges        6531
Churn                  2
dtype: int64

In [43]:
df.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [44]:
df.duplicated().sum()

0

In [45]:
df.isna().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [46]:
pd.to_numeric(df['TotalCharges'], errors='coerce')

0         29.85
1       1889.50
2        108.15
3       1840.75
4        151.65
         ...   
7038    1990.50
7039    7362.90
7040     346.45
7041     306.60
7042    6844.50
Name: TotalCharges, Length: 7043, dtype: float64

# Find customers who are at high risk of churn with the criteria that customers must have a monthly contract and their monthly bill is above the 90th percentile.

In [47]:
df['MonthlyCharges'].value_counts()

MonthlyCharges
20.05     61
19.85     45
19.95     44
19.90     44
20.00     43
          ..
23.65      1
114.70     1
43.65      1
87.80      1
78.70      1
Name: count, Length: 1585, dtype: int64

In [48]:
df['Contract']

0       Month-to-month
1             One year
2       Month-to-month
3             One year
4       Month-to-month
             ...      
7038          One year
7039          One year
7040    Month-to-month
7041    Month-to-month
7042          Two year
Name: Contract, Length: 7043, dtype: object

In [49]:
df.mcpersentil90 = np.percentile(df['MonthlyCharges'], 90)
condition = (df['Contract'] == 'Month-to-month') & (df['MonthlyCharges'] > df.mcpersentil90)
df.pelangganrisikotinggi = df[condition]
df.pelangganrisikotinggi

<ipython-input-49-58dd799dd232>:3: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
  df.pelangganrisikotinggi = df[condition]


customerID  gender  SeniorCitizen Partner Dependents  tenure  \
8     7892-POOKP  Female              0     Yes         No      28   
13    0280-XJGEX    Male              0      No         No      49   
14    5129-JLPIS    Male              0      No         No      25   
38    5380-WJKOV    Male              0      No         No      34   
83    9919-YLNNG  Female              0      No         No      42   
...          ...     ...            ...     ...        ...     ...   
6914  7142-HVGBG    Male              1     Yes         No      43   
6952  6394-MFYNG  Female              0      No         No      48   
7006  0093-XWZFY    Male              0      No         No      40   
7023  1035-IPQPU  Female              1     Yes         No      63   
7034  0639-TSIQW  Female              0      No         No      67   

     PhoneService MultipleLines InternetService OnlineSecurity  ...  \
8             Yes           Yes     Fiber optic             No  ...   
13            Yes           Yes     Fiber optic             No  ...   
14            Yes            No     Fiber optic            Yes  ...   
38            Yes           Yes     Fiber optic             No  ...   
83            Yes            No     Fiber optic             No  ...   
...           ...           ...             ...            ...  ...   
6914          Yes           Yes     Fiber optic             No  ...   
6952          Yes           Yes     Fiber optic             No  ...   
7006          Yes           Yes     Fiber optic             No  ...   
7023          Yes           Yes     Fiber optic             No  ...   
7034          Yes           Yes     Fiber optic            Yes  ...   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
8                 Yes         Yes         Yes             Yes  Month-to-month   
13                Yes          No         Yes             Yes  Month-to-month   
14                Yes         Yes         Yes             Yes  Month-to-month   
38                Yes          No         Yes             Yes  Month-to-month   
83                Yes         Yes         Yes             Yes  Month-to-month   
...               ...         ...         ...             ...             ...   
6914              Yes          No         Yes             Yes  Month-to-month   
6952              Yes          No         Yes             Yes  Month-to-month   
7006              Yes          No         Yes             Yes  Month-to-month   
7023              Yes          No         Yes             Yes  Month-to-month   
7034              Yes          No         Yes              No  Month-to-month   

     PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
8                 Yes           Electronic check         104.80       3046.05   
13                Yes  Bank transfer (automatic)         103.70        5036.3   
14                Yes           Electronic check         105.50       2686.05   
38                Yes           Electronic check         106.35       3549.25   
83                Yes  Bank transfer (automatic)         103.80        4327.5   
...               ...                        ...            ...           ...   
6914              Yes           Electronic check         103.00        4414.3   
6952              Yes  Bank transfer (automatic)         107.40        5121.3   
7006              Yes    Credit card (automatic)         104.50       4036.85   
7023              Yes           Electronic check         103.50        6479.4   
7034              Yes    Credit card (automatic)         102.95       6886.25   

     Churn  
8      Yes  
13     Yes  
14      No  
38     Yes  
83      No  
...    ...  
6914   Yes  
6952   Yes  
7006   Yes  
7023    No  
7034   Yes  

[207 rows x 21 columns]

The 207 customers in the high-risk category, may be dissatisfied with the service or perceive the cost as too high.

# Grouping data by contract type and calculating average monthly billing and churn percentage for each group.

In [50]:
df['Contract'].value_counts()

Contract
Month-to-month    3875
Two year          1695
One year          1473
Name: count, dtype: int64

In [51]:
df['Churn'] = df['Churn'].map({'Yes':1, 'No':0})

In [52]:
df['Churn'].head()

0    0
1    0
2    1
3    0
4    1
Name: Churn, dtype: int64

In [56]:
df.contractgroup = df.groupby('Contract').agg({'MonthlyCharges':'mean', 'Churn' : 'mean'})
df.contractgroup['Churn (%)'] = df.contractgroup['Churn'] * 100
df.contractgroup

MonthlyCharges     Churn  Churn (%)
Contract                                           
Month-to-month       66.398490  0.427097  42.709677
One year             65.048608  0.112695  11.269518
Two year             60.770413  0.028319   2.831858

Customers with month-to-month contracts have the highest churn percentage (42.71%) compared to customers with one or two year contracts.

# Pivot table to analyze the relationship between contract type and payment method on average monthly billing.

In [57]:
pd.pivot_table(
    data = df,
    index = 'Contract',
    columns = 'PaymentMethod',
    values = 'MonthlyCharges',
    aggfunc = 'mean'
)

PaymentMethod   Bank transfer (automatic)  Credit card (automatic)  \
Contract                                                             
Month-to-month                  69.085399                67.678821   
One year                        67.484143                67.972362   
Two year                        65.013918                64.422117   

PaymentMethod   Electronic check  Mailed check  
Contract                                        
Month-to-month         74.989486     46.050056  
One year               79.095245     44.306380  
Two year               84.335714     38.587304

Annual contracts have lower average monthly bills than monthly contracts. Electronic checks have the highest average bills, while mailed checks have the lowest for all contract types.

# Crosstab to evaluate the relationship between churn and payment method, and calculate the churn percentage for each payment method.

In [58]:
df.churnmethod = pd.crosstab(df['PaymentMethod'],df['Churn'])
df.churnmethod['Churn (%)'] = df.churnmethod[1]/(df.churnmethod[0]+df.churnmethod[1])*100
df.churnmethod

<ipython-input-58-63c6c37c0f26>:1: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
  df.churnmethod = pd.crosstab(df['PaymentMethod'],df['Churn'])


Churn                         0     1  Churn (%)
PaymentMethod                                   
Bank transfer (automatic)  1286   258  16.709845
Credit card (automatic)    1290   232  15.243101
Electronic check           1294  1071  45.285412
Mailed check               1304   308  19.106700

Payment method has a significant relationship with churn. Customers with Electronic check payment method showed the highest churn percentage compared to other payment methods.

# Aggregation to evaluate total churn with senior citizen yes and dependents more than 0.

In [59]:
df['Dependents'] = df['Dependents'].map({'Yes': 1, 'No': 0})
condition = (df['SeniorCitizen'] == 1) & (df['Dependents'] > 0)
df.condi = df[condition]
df.condi

<ipython-input-59-b80ae5b0d6ac>:3: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
  df.condi = df[condition]


customerID  gender  SeniorCitizen Partner  Dependents  tenure  \
54    4667-QONEA  Female              1     Yes           1      60   
57    5067-XJQFU    Male              1     Yes           1      66   
72    1891-QRQSA    Male              1     Yes           1      64   
140   3146-MSEGF  Female              1     Yes           1      72   
308   3167-SNQPL    Male              1     Yes           1      38   
...          ...     ...            ...     ...         ...     ...   
6235  5180-UCIIQ    Male              1     Yes           1      22   
6346  9661-MHUMO    Male              1     Yes           1       1   
6446  2205-LPVGL    Male              1     Yes           1      58   
6451  4868-AADLV    Male              1     Yes           1      66   
6763  7471-MQPOS    Male              1     Yes           1      71   

     PhoneService     MultipleLines InternetService OnlineSecurity  ...  \
54            Yes                No             DSL            Yes  ...   
57            Yes               Yes     Fiber optic             No  ...   
72            Yes               Yes     Fiber optic            Yes  ...   
140           Yes               Yes             DSL            Yes  ...   
308           Yes               Yes     Fiber optic             No  ...   
...           ...               ...             ...            ...  ...   
6235           No  No phone service             DSL            Yes  ...   
6346          Yes                No             DSL             No  ...   
6446          Yes               Yes     Fiber optic            Yes  ...   
6451          Yes               Yes     Fiber optic            Yes  ...   
6763          Yes               Yes             DSL             No  ...   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
54                Yes         Yes          No             Yes        One year   
57                Yes         Yes         Yes             Yes        One year   
72                Yes         Yes         Yes             Yes        Two year   
140               Yes         Yes         Yes             Yes        Two year   
308                No          No         Yes             Yes  Month-to-month   
...               ...         ...         ...             ...             ...   
6235               No          No          No             Yes  Month-to-month   
6346               No         Yes          No              No  Month-to-month   
6446               No         Yes          No              No        One year   
6451              Yes         Yes         Yes             Yes        One year   
6763              Yes         Yes          No             Yes        One year   

     PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
54                Yes    Credit card (automatic)          74.85       4456.35   
57                Yes           Electronic check         108.45       7076.35   
72                Yes  Bank transfer (automatic)         111.60          7099   
140               Yes    Credit card (automatic)          88.05       6425.65   
308                No           Electronic check         101.15       3741.85   
...               ...                        ...            ...           ...   
6235               No               Mailed check          40.05         880.2   
6346              Yes    Credit card (automatic)          49.25         49.25   
6446              Yes  Bank transfer (automatic)          89.85       5125.75   
6451              Yes           Electronic check         116.25       7862.25   
6763              Yes  Bank transfer (automatic)          72.90       5139.65   

     Churn  
54       0  
57       0  
72       0  
140      0  
308      0  
...    ...  
6235     1  
6346     1  
6446     0  
6451     0  
6763     0  

[91 rows x 21 columns]

There are 91 out of 7043 customers who are Senior Citizens with dependents > 0 and experienced churn. This shows that this group has a small contribution to the total churn as a whole.